src/mlang/m_frontend/mast_to_mir.ml

```diff
@@ -454,10 +454,175 @@ let get_constants (p : Mast.program) : float Pos.marked ConstMap.t =
 let belongs_to_iliad_app (r : Mast.application Pos.marked list) : bool =
   List.exists (fun app -> Pos.unmark app = "iliad") r
 
+let sort_attributes (attrs : Mast.variable_attribute list) =
+  List.sort
+    (fun c1 c2 -> String.compare (Pos.unmark (fst c1)) (Pos.unmark (fst c2)))
+    attrs
+
+let get_var_categories (p : Mast.program) =
+  let categories =
+    List.fold_left
+      (fun decls source_file ->
+        List.fold_left
+          (fun decls source_file_item ->
+            match Pos.unmark source_file_item with
+            | Mast.VarCatDecl (catdecl, pos) ->
+                let normalized_decl =
+                  {
+                    catdecl with
+                    var_category =
+                      List.sort
+                        (fun c1 c2 ->
+                          String.compare (Pos.unmark c1) (Pos.unmark c2))
+                        catdecl.var_category;
+                    var_attributes =
+                      List.sort
+                        (fun c1 c2 ->
+                          String.compare (Pos.unmark c1) (Pos.unmark c2))
+                        catdecl.var_attributes;
+                  }
+                in
+                let already_defined =
+                  let decl_l = List.length normalized_decl.var_category in
+                  List.find_opt
+                    (fun (decl, _pos) ->
+                      decl_l = List.length decl.Mast.var_category
+                      && List.for_all2
+                           (fun a b ->
+                             String.equal (Pos.unmark a) (Pos.unmark b))
+                           normalized_decl.var_category decl.Mast.var_category)
+                    decls
+                in
+                begin
+                  match already_defined with
+                  | None -> ()
+                  | Some (_decl, pos) ->
+                      Cli.warning_print
+                        "Category \"%s\" defined more than once:@;\
+                         Already defined %a"
+                        (String.concat " "
+                           (Format_mast.format_var_type normalized_decl.var_type
+                           :: List.map Pos.unmark normalized_decl.var_category))
+                        Pos.format_position pos
+                end;
+                (normalized_decl, pos) :: decls
+            | _ -> decls)
+          decls source_file)
+      [] p
+  in
+  let categories =
+    (* Sorted to match longest category first *)
+    List.sort
+      (fun c1 c2 ->
+        compare
+          (List.length (Pos.unmark c2).Mast.var_category)
+          (List.length (Pos.unmark c1).Mast.var_category))
+      categories
+  in
+  if categories = [] then
+    Cli.warning_print
+      "No variable categories defined. No check will be performed.";
+  categories
+
+let check_var_category (categories : Mast.var_category_decl Pos.marked list)
```

Comment:
Yes, although I am not sure of what relaxing you meant exactly.

Reply:
In the behavior introduced by this PR, if you declare a variable of category `X` with all the expected attributes for `X` AND SOME MORE, then it's OK to have more attributes, right?

Reply:
It's ok to have more category keywords, not more attributes. Say a category `X Y`, a variable of category `X Y Z` would fit, but a warning would be raised if the sets of attribute are not the same.

And in fact it should be an error, since those attributes will eventually end up as field of a C struct in the generated code. We can't allow variables that would need more fields than declared.